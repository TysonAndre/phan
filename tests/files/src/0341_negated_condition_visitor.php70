<?php

/**
 * @param int|string $x
 * @param int|string $y
 */
function testNegate341($x, $y) {
    if (!is_string($x)) {
        intdiv($x, 2);
    } else {
        $z = strlen($x);
    }

    if (!is_string($y)) {
        echo strlen($y);
    } else {
        echo intdiv($y, 2);
    }
}
/** @param ?int $x */
function testNegate341B($x) {
    if (!isset($x)) {
        intdiv($x, 2);  // wrong.
    } else if (rand() % 2 === 1) {
        // do nothing
    } else {
        echo strlen($x);  // should infer $x is int and warn
    }
}
/** @param ?int $x */
function testNegate341C($x) {
    if (empty($x)) {
        intdiv($x, 2);  // wrong if it's null, can be null or 0 (?int)
    } else {
        echo strlen($x);  // should infer $x is int and warn
    }
}

/**
 * @param int|string $x
 */
function testNegate341D($x) {
    return is_string($x)
        ? intdiv($x, 2)  // wrong, a string
        : strlen($x);  // wrong, an int
}

/**
 * @param int|string $x
 */
function testNegate341E($x) {
    is_string($x)
        ? intdiv($x, 2)  // wrong, a string
        : strlen($x);  // wrong, an int
}
