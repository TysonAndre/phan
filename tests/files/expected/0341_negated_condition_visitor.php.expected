%s:11 PhanUnusedVariable Unused definition of variable $z
%s:15 PhanTypeMismatchArgumentInternal Argument 1 ($string) is $y of type int but \strlen() takes string
%s:17 PhanTypeMismatchArgumentInternal Argument 1 ($num1) is $y of type string but \intdiv() takes int
%s:23 PhanTypeMismatchArgumentInternal%SReal Argument 1 ($num1) is $x of type null but \intdiv() takes int
%s:27 PhanTypeMismatchArgumentInternal Argument 1 ($string) is $x of type int but \strlen() takes string
%s:33 PhanTypeMismatchArgumentNullableInternal Argument 1 ($num1) is $x of type ?0 but \intdiv() takes int (expected type to be non-nullable)
%s:35 PhanTypeMismatchArgumentInternal Argument 1 ($string) is $x of type non-zero-int but \strlen() takes string
%s:44 PhanTypeMismatchArgumentInternal Argument 1 ($num1) is $x of type string but \intdiv() takes int
%s:45 PhanTypeMismatchArgumentInternal Argument 1 ($string) is $x of type int but \strlen() takes string
%s:52 PhanNoopTernary Unused result of a ternary expression where the true/false results don't seem to have side effects
%s:53 PhanTypeMismatchArgumentInternal Argument 1 ($num1) is $x of type string but \intdiv() takes int
%s:54 PhanTypeMismatchArgumentInternal Argument 1 ($string) is $x of type int but \strlen() takes string
